### ISSUE TITLE
plugin install fails on Calibre5

### ISSUE THREAD

**Issue by hamkg** [apprenticeharper/DeDRM_tools#1245]
After upgrading to Calibre-5.0.0 the plugin couldn't be found.
When trying to install it I get a syntax error for DeDRM__init__py line 168 
                print u"{0} v{1}: Copying needed library files from plugin's zip".format(PLUGIN_NAME, PLUGIN_VERSION)

Reverted to Calibre 4.23, plugin found and worked as before.

**Comment by Myuki** [comment]
> 
> 
> After upgrading to Calibre-5.0.0 the plugin couldn't be found.
> When trying to install it I get a syntax error for DeDRM__init__py line 168
> print u"{0} v{1}: Copying needed library files from plugin's zip".format(PLUGIN_NAME, PLUGIN_VERSION)
> 
> Reverted to Calibre 4.23, plugin found and worked as before.

Calibre 5 has migrated from Python 2 to Python 3. DeDRM is still written in Python 2 and need to be translated or rewrote with Python 3.

**Comment by quantumchaos451** [comment]
This patch allows me to load the DeDRM plugin in Calibre 5. Still needs more testing though.
[python3_dedrm.patch.txt](https://github.com/apprenticeharper/DeDRM_tools/files/5281690/python3_dedrm.patch.txt)

**Comment by protist** [comment]
There's been a bit of work in issue #979, including a semi-working python 3 port [here](https://github.com/lalmeras/DeDRM_tools/tree/Python3).

**Comment by quantumchaos451** [comment]
Cool. OK. Will have a look at that port.

**Comment by hamkg** [comment]
Thanks for the input, have tried the suggestions without success. Ill stick with Calibre-4.23 for now and hope that ApprenticeHarper and other contributors will come up with a solution in due course.

**Comment by protist** [comment]
@hamkg Yes, I also couldn't get the fork working either. I posted in the other issue. My understanding is that ApprenticeHarper is missing unfortunately, and appears unlikely to update this repository.

**Comment by ElleKayEm** [comment]
Apprentice Harper hasn't had much time for this in years, but the last update was in June this year.  I wouldn't give up hope for a reappearance yet.  Now that Python 3 calibre is officially released, I imagine either Harper will be back or others will continue this work.  Up to now, it was mostly just effecting Linux calibre users.

**Comment by hamkg** [comment]
My thoughts as well. It might be that the stable Calibre-5 contains other changes, causing issues with the forks (were most commits are a few months old).

**Comment by kdurnin** [comment]
I am having the same problem on Calibre 5.0.1. The 6.8.0 DRM zip for Calibre will not install
[Cannot install DRM 680 with Calibre501.txt](https://github.com/apprenticeharper/DeDRM_tools/files/5285448/Cannot.install.DRM.680.with.Calibre501.txt)

**Comment by ElleKayEm** [comment]
You will need to stick with calibre 4.23 to install DeDRM 6.8.0.

**Comment by kdurnin** [comment]
Thank you!!! Really appreciate your help.Sent from my T-Mobile 4G LTE Device
-------- Original message --------From: ElleKayEm <notifications@github.com> Date: 9/25/20  4:59 PM  (GMT-06:00) To: apprenticeharper/DeDRM_tools <DeDRM_tools@noreply.github.com> Cc: kdurnin <kdurnin@verizon.net>, Comment <comment@noreply.github.com> Subject: Re: [apprenticeharper/DeDRM_tools] plugin install fails on Calibre5
  (#1245) 
You will need to stick with calibre 4.23 to install DeDRM 6.8.0.

**Comment by Kaltininkas** [comment]
I am also having problems...here's a copy of the error I received with the log:

calibre, version 5.0.1
ERROR: Unhandled exception: <b>SyntaxError</b>:invalid syntax (calibre_plugins.dedrm.__init__, line 168)

calibre 5.0.1 [64bit]  embedded-python: True is64bit: True
Windows-10-10.0.19041-SP0 Windows ('64bit', 'WindowsPE')
('Windows', '10', '10.0.19041')
Python 3.8.5
Windows: ('10', '10.0.19041', 'SP0', 'Multiprocessor Free')
Interface language: None
Successfully initialized third party plugins: Obok DeDRM (6, 7, 0)
Traceback (most recent call last):
  File "site-packages\calibre\gui2\preferences\plugins.py", line 317, in add_plugin
  File "site-packages\calibre\customize\ui.py", line 472, in add_plugin
  File "site-packages\calibre\customize\ui.py", line 61, in load_plugin
  File "site-packages\calibre\customize\zipplugin.py", line 200, in load
  File "site-packages\polyglot\builtins.py", line 123, in reload
  File "importlib\__init__.py", line 169, in reload
  File "<frozen importlib._bootstrap>", line 602, in _exec
  File "site-packages\calibre\customize\zipplugin.py", line 179, in load_module
  File "calibre_plugins.dedrm.__init__", line 168
    print u"{0} v{1}: Copying needed library files from plugin's zip".format(PLUGIN_NAME, PLUGIN_VERSION)
          ^
SyntaxError: invalid syntax

**Comment by ElleKayEm** [comment]
You will need to use calibre 4.23 until DeDRM gets updated to Python 3 compatibility.

**Comment by mejoyus** [comment]
thanks for the info re calibre 5.0.1 ive reverted back .... can any else tell me if they have now found they can not now use the old link for kindle as it now wont register ... if i download the newest app onto my pc will register fine but even though ive added the KFX imput and have the DeDRM showing it will not let me convert as file says zip when import and if i try to convert says DRM protected ... thanks for any help

**Comment by protist** [comment]
> As long as one keeps Python2 aboard, it's also possible to use DeDRM from the command line as described here:
> https://news.ycombinator.com/item?id=24589822
> I tried this with a DRM encrypted .epub file, worked perfectly.

Thanks @hamkg . I'd be totally happy to skip the Calibre integration, as long as I can get it working! 

Could you please tell me where you found your old key? I can't find anything in my system with `k4i` or `.der` in it

**Comment by hamkg** [comment]
@protist,
can't remember how I (years ago) found my adeptkey.der and (wisely) stored it for future use (e.g. when migrating to a new pc) but I probably followed a guide from somewhere along the lines of https://expertvista.com/2011/remove-drm-howto/
It might be that I had to do it on a windows system although I otherwise run Linux (with ADE4 under wine).

**Comment by jerome-diver** [comment]
@protist you can use the command "find" of linux system to find any file on your system file.
```
find /my_source/directory/i_m_going_to_search_in -name "*.k4i" -type f
```
this will show you a list of file inside your directory source to search in recursively of files named by everything extended by .k4i
if you find nothing, it is because there is nothing.

**Comment by protist** [comment]
@hamkg Ech okay, thank you for the link. I'm also on Linux, so I can try it with Wine, but no high hopes. I'll have a look for other alternatives if it doesn't work.

@jerome-diver Thanks, I probably should have been more specific. I just used (m)`locate` instead to search my entire indexed file system instead. Much much faster and more comprehensive! Thank you anyway.

**Comment by ElleKayEm** [comment]
@mejoyus 
For KFX files, you need version 1.26 or earlier of the Kindle app.  Or disable KFX downloads.  See this post:  https://www.mobileread.com/forums/showthread.php?t=283371

**Comment by stainesa** [comment]
Same issues Error is

Will go back a version of Calibre.
calibre, version 5.0.1
ERROR: Unhandled exception: <b>SyntaxError</b>:invalid syntax (calibre_plugins.dedrm.__init__, line 168)

calibre 5.0.1  embedded-python: True is64bit: False
Windows-10-10.0.18362-SP0 Windows ('32bit', 'WindowsPE')
32bit process running on 64bit windows
('Windows', '10', '10.0.18362')
Python 3.8.5
Windows: ('10', '10.0.18362', 'SP0', 'Multiprocessor Free')
Interface language: None
Successfully initialized third party plugins: 
Traceback (most recent call last):
  File "site-packages\calibre\gui2\preferences\plugins.py", line 317, in add_plugin
  File "site-packages\calibre\customize\ui.py", line 472, in add_plugin
  File "site-packages\calibre\customize\ui.py", line 61, in load_plugin
  File "site-packages\calibre\customize\zipplugin.py", line 200, in load
  File "site-packages\polyglot\builtins.py", line 123, in reload
  File "importlib\__init__.py", line 169, in reload
  File "<frozen importlib._bootstrap>", line 602, in _exec
  File "site-packages\calibre\customize\zipplugin.py", line 179, in load_module
  File "calibre_plugins.dedrm.__init__", line 168
    print u"{0} v{1}: Copying needed library files from plugin's zip".format(PLUGIN_NAME, PLUGIN_VERSION)
          ^
SyntaxError: invalid syntax

**Comment by ElleKayEm** [comment]
@stainesa 
For now, you will need to use calibre version 4.23 in order to use DeDRM.  Calibre 5 has Python 3 and DeDRM is not compatible yet.

**Comment by mejoyus** [comment]
> @mejoyus
> For KFX files, you need version 1.26 or earlier of the Kindle app. Or disable KFX downloads. See this post: https://www.mobileread.com/forums/showthread.php?t=283371

thanks .. thats my problem the earlier version of kindle wont now register ... looks like amazon have closed it? will have a look at the link to disable KFX

**Comment by ElleKayEm** [comment]
I seen reports from others with that issue.  They were able to fix this by doing a complete deregister and uninstall of the Kindle app and associated files.  Then they were able to install and register an older version.

**Comment by mejoyus** [comment]
> I seen reports from others with that issue. They were able to fix this by doing a complete deregister and uninstall of the Kindle app and associated files. Then they were able to install and register an older version.

ahh i did remove it but i didnt remove the files .....

**Comment by mejoyus** [comment]
was wondering if getting an old kindle ereader was worth it as think the older ones you can plug into your computer and download the files that way

**Comment by ElleKayEm** [comment]
If you have an E-ink Kindle on your account, you can download your books from the website.  Then you enter the Kindle's serial number into the DeDRM plugin customization, and those books can be decrypted.  With older Kindles which don't get updates anymore (not sure which models), you'd be able to DeDRM books transfered directly from the device via USB.

**Comment by mejoyus** [comment]
> If you have an E-ink Kindle on your account, you can download your books from the website. Then you enter the Kindle's serial number into the DeDRM plugin customization, and those books can be decrypted. With older Kindles which don't get updates anymore (not sure which models), you'd be able to DeDRM books transfered directly from the device via USB.

thanks thats very helpful :)

**Comment by apprenticeharper** [comment]
I am trying to find time to merge code, check it works, and get an 'official' release out.

**Comment by mejoyus** [comment]
> @mejoyus
> For KFX files, you need version 1.26 or earlier of the Kindle app. Or disable KFX downloads. See this post: https://www.mobileread.com/forums/showthread.php?t=283371

sorry to bother you again can you tell me where i can get link 1.26 as the 1.17 just wont let me register keeps saying wrong email nothing showing under amazon devices so nothing to deregister as it wont let me register and removed and reloaded a few times ... found some of the links but seems to skip from 1.17 (2016) to ones late last year till current and im assuming those will be to new ...

**Comment by hrmcrm** [comment]
> I am trying to find time to merge code, check it works, and get an 'official' release out.

Thank you for putting in your time.

**Comment by hugohamelcom** [comment]
Getting this error:
calibre, version 5.0.1
ERROR: Unhandled exception: <b>SyntaxError</b>:invalid syntax (calibre_plugins.dedrm.__init__, line 168)

calibre 5.0.1 Portable embedded-python: True is64bit: False
Windows-10-10.0.18362-SP0 Windows ('32bit', 'WindowsPE')
32bit process running on 64bit windows
('Windows', '10', '10.0.18362')
Python 3.8.5
Windows: ('10', '10.0.18362', 'SP0', 'Multiprocessor Free')
Interface language: None
Successfully initialized third party plugins: Obok DeDRM (6, 7, 0)
Traceback (most recent call last):
  File "site-packages\calibre\gui2\preferences\plugins.py", line 317, in add_plugin
  File "site-packages\calibre\customize\ui.py", line 472, in add_plugin
  File "site-packages\calibre\customize\ui.py", line 61, in load_plugin
  File "site-packages\calibre\customize\zipplugin.py", line 200, in load
  File "site-packages\polyglot\builtins.py", line 123, in reload
  File "importlib\__init__.py", line 169, in reload
  File "<frozen importlib._bootstrap>", line 602, in _exec
  File "site-packages\calibre\customize\zipplugin.py", line 179, in load_module
  File "calibre_plugins.dedrm.__init__", line 168
    print u"{0} v{1}: Copying needed library files from plugin's zip".format(PLUGIN_NAME, PLUGIN_VERSION)
          ^
SyntaxError: invalid syntax

**Comment by kau78** [comment]
> I am trying to find time to merge code, check it works, and get an 'official' release out.

thx!!

**Comment by mejoyus** [comment]
hi ...  was having issues with my old kindle app wouldnt let me register kept saying not recognized email ... so got an old kindle ereader which is downloading the old style azw file loaded a couple of books over to calibre this way yesterday went to move another today and its moving over but wont let me convert to epub says DRM locked .. ive checked and updated to latest DeDRM 6.8 but has made no difference .. should i be using an older version or can anyone show me what i can do to sort .. many thanks

**Comment by mejoyus** [comment]
> > If you have an E-ink Kindle on your account, you can download your books from the website. Then you enter the Kindle's serial number into the DeDRM plugin customization, and those books can be decrypted. With older Kindles which don't get updates anymore (not sure which models), you'd be able to DeDRM books transfered directly from the device via USB.
> 
> thanks thats very helpful :)

ahh see i need the serial number

**Comment by ElleKayEm** [comment]
@mejoyus 
Make sure you have the serial number correct.  If you're still having trouble, start your own issue and post a log.

**Comment by mejoyus** [comment]
> @mejoyus
> Make sure you have the serial number correct. If you're still having trouble, start your own issue and post a log.

its an old touch 2012 ish but tried the serial number but will try again do i leave the spaces in and will try O and 0 just in case i got them mixed

**Comment by mejoyus** [comment]
> @mejoyus
> Make sure you have the serial number correct. If you're still having trouble, start your own issue and post a log.

many thanks serial number did it i had put O instead of 0 for one of them

**Comment by tartley** [comment]
I think this issue can be closed for now, as a dupe of https://github.com/apprenticeharper/DeDRM_tools/issues/1324. Shout if you disagree.

**Comment by mu22le** [comment]
You usually close the more recent issue in favor of the older one, which, as in this case, contains more information. 

So #1324 should be closed, not #1245. Or maybe both should be closed in favor of #979.

Finally, if you want to close #1245 at least clearly mark it as a duplicate, so people will see it as soon as they open the page.

**Comment by tartley** [comment]
Acknowledged, thanks. I had wrongly assumed that, since nobody had been doing this, no-one cared, and since this problem was already so well understood, it didn't matter which issue was treated as the canonical one. But I appreciate your guidance, and will fix when I'm at a real keyboard.

**Comment by tartley** [comment]
@mu22le Hey. I had a look in detail at which specific issues you linked to.

I don't think it's the most constructive to close issues in favor of #979. That issue has really big scope. It's hard to understand all the details of, and it's difficult to co-ordinate several people working on it in parallel, since if one person says "I'm working on #979", then this by itself doesn't give anyone else any information about what they ought to do to avoid duplicating or conflicting work. They need extra out of band conversations to have that discussion. As a result of all the above, the issue has been very long lasting, even though there is a large amount of interest and activity and work around it, and it is still nowhere near getting closed. Good issues are _actionable_. That means smaller scope - something an individual can bite off, and complete, and mark as "done" in a short amount of time - hours or days. Even though it has a long conversation attached to it, closing issues as dupes of #979 would provide contributors with substantially less information and guidance about what they ought to work on, and how the collaborate on those goals, not more.

I guess what I'm saying is, I'd rather keep #1245 as a representative of the dupe "failure into install the plugin" issues. (rather than #979) Does that sound good? Or am I missing something?

Sincere thanks for your guidance and opinions. I'm not aiming to argue, I just want to understand what y'all think, then I will do whatever your consensus is.

**Comment by tartley** [comment]
@mu22le: I took the liberty of re-opening this, and closing https://github.com/apprenticeharper/DeDRM_tools/issues/1324 as a dupe of this, as you suggest.

>  at least clearly mark it as a duplicate of another bug

I've been attaching the label 'duplicate', and commenting with a link to the still-open dupe when I close the ticket. Do people think something else is needed?

**Comment by wienfuchs** [comment]
Same problem here with both Cakibre 5.5 an new 5.6.0 version, alos on new releas of 6.8.1. Just to mention: This was addressed in the comments of the [release page](https://github.com/apprenticeharper/DeDRM_tools/releases) and probably will be solved in release 7.0 as far as I read. At the moment I did not need it, so I personally can wait- others may have a look into 7.0.0.b3 Beta or later.

**Comment by ElleKayEm** [comment]
For calibre 5 you can only install the beta DeDRM 7 pre-release.  If you want to install DeDRM 6.8.x, you need to use calibre 4.23 or earlier.

**Comment by wienfuchs** [comment]
> For calibre 5 you can only install the beta DeDRM 7 pre-release. If you want to install DeDRM 6.8.x, you need to use calibre 4.23 or earlier.

That's what I wanted to express ;)

**Comment by ElleKayEm** [comment]
You just said it like this was unexpected.  DeDRM 7 is the rewrite to Python 3 required for calibre 5 plugins.  DeDRM 6.8 is in Python 2 and calibre 4.23 is the last version which it will work with.

**Comment by wienfuchs** [comment]
> You just said it like this was unexpected. DeDRM 7 is the rewrite to Python 3 required for calibre 5 plugins. DeDRM 6.8 is in Python 2 and calibre 4.23 is the last version which it will work with.

In fact to me it was a bit unexpected as I did not realise DeDRM was "gone" already bye release of calibre 5.x - i got the release 6.8.1 notification today, where there was no note of this in the mail and just downloaded it and it was throwing an installation error. Only then I got further and found this issue - and only wanted to mention the already known circumstanced for other noobish non-followers like me to find an easy "status". Nothing more, nothing less. I very much appreciate the good work and just wanted to help others finding this information.

**Comment by ElleKayEm** [comment]
I just find that odd is all.  The email I received included the release notes and so does the page you download it from.  Seems like a more likely place to come across the info than this old issue.

**Comment by wienfuchs** [comment]
> The email I received included the release notes 

But no info, that it is not working with Calibre 5.x
![grafik](https://user-images.githubusercontent.com/44177781/100526042-bb8f9680-31c5-11eb-9f00-40d6f07e57e1.png)

> and so does the page you download it from. 

There indeed it shows "This release is for calibre 4.x and earlier". But this often is not read by people, and I did not read it there too.

> Seems like a more likely place to come across the info than this old issue.

As I only found out there might or is an information whe trying to install 6.8.1 on Calibre 5.5.x and 5.6.0 and then searched for it: Maybe you are right, but finally I landed here and so maybe others too. What is wrong with a simple status info for others I added so that this now had to lead to this odd discussion? I will stop discussing that at this point, my intention was just slight info, it was placed:

> Same problem here with both Calibre 5.5 an new 5.6.0 version, also by using new release of 6.8.1.
> Just to mention: This fact was addressed in the comments of the [release page](https://github.com/apprenticeharper/DeDRM_tools/releases) and probably will be solved in release 7.0 as far as I read. At the moment I did not need it, so I personally can wait- others may have a look into 7.0.0.b3 Beta or later.

Nothing more to say in fact.

**Comment by ElleKayEm** [comment]
Just trying to understand.  Wondering if you had come from somewhere where the release had been reposted without the notes.
